scipio/src/sys/uring.rs

```diff
@@ -215,15 +215,13 @@ where
             return Some(());
         }
 
-        // Will be None if it was cancelled while in-flight.
-        if let Some(src) = consume_source(value.user_data()) {
-            let source = mut_source(&src);
-
-            if try_process(source).is_none() {
-                let mut w = source.wakers.borrow_mut();
-                w.result = Some(value.result());
-                wakers.append(&mut w.waiters);
-            }
+        let src = consume_source(value.user_data());
```

Comment:
In general I like the idea of pairing operations. It is easier to get it right and more robust.

However note that both those operations are not the same (we'll now discuss whether or not that matters, but...)

Prior to your patch we would only push a waker if the operation was *not* cancelled, because `try_process` was inside the `consume_source` branch.

Now we may be queueing a waker for an operation that is cancelled which may have other nasty side effects too.

Whether or not the operation was cancelled is encoded in the `result`, so we can still have it both ways: we can require consume, the way you do, but test for the `result` before calling `try_process`.

Reply:
Hm, yeah, agree that this was an unintended change in the behavior, I've put the old behavior back in place.

However, it does seem surprising and lost-wakeupy to me. Intuitively, I would assume that, when we cancel an op, we need to wake everyone waiting on op and notify them that their waiting is now futile. But, even if that reasoning is true, it should be dealt with in a separate PR.

Reply:
You may be right and I think a great way of demonstrating that would be through testing those conditions specifically.
The reason this is done this way is essentially driven by the fact that cancellations right now are mostly coming from cancelled tasks in which case there is no one around to receive the wake.

Whether they will be ignored or leak in that situation is speculative for now. I am putting a lot of faith in my new friend miri to help us figure that out as we add more tests for those cases